File: docker/pi_stream.py
"""Parse Pi CLI --mode json output into structured harness events."""

from __future__ import annotations

import json
from typing import Any, Generator, Iterator

from claude_stream import _emit_image_events


def parse_pi_stream(lines: Iterator[str]) -> Generator[dict[str, Any], None, None]:
    parser = new_pi_stream_parser()
    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            continue
        yield from parser.handle(obj)


def new_pi_stream_parser() -> "_PiStreamParser":
    return _PiStreamParser()


class _PiStreamParser:
    def __init__(self) -> None:
        self.emitted_text = False
        self.needs_text_sep = False
        self.seen_tool_starts: set[str] = set()
        self.seen_tool_ends: set[str] = set()
        self.seen_tool_results: set[str] = set()

    def _mark_text_sep_needed(self) -> None:
        if self.emitted_text:
            self.needs_text_sep = True

    def _emit_text(self, text: str) -> Generator[dict[str, Any], None, None]:
        if not text:
            return
        if self.emitted_text and self.needs_text_sep and not text.startswith("\n"):
            text = "\n\n" + text
        self.needs_text_sep = False
        self.emitted_text = True
        yield {"type": "text", "content": text}

    def handle(self, obj: dict[str, Any]) -> Generator[dict[str, Any], None, None]:
        t = obj.get("type")

        if t == "session":
            sid = obj.get("id") or ""
            if sid:
                yield {"type": "session_id", "sessionId": sid}
            return

        if t == "message_update":
            yield from self._handle_message_update(obj)
            return

        if t == "tool_execution_start":
            self._mark_text_sep_needed()
            tool_id = obj.get("toolCallId") or ""
            tool_name = obj.get("toolName") or ""
            if tool_id and tool_id not in self.seen_tool_starts:
                self.seen_tool_starts.add(tool_id)
                yield {"type": "tool_call_start", "toolCallId": tool_id, "toolName": tool_name}
            return

        if t == "tool_execution_end":
            yield from self._emit_tool(
                obj.get("toolCallId") or "",
                obj.get("toolName") or "",
                obj.get("args"),
                obj.get("result"),
            )
            return

        if t == "turn_end":
            for result in obj.get("toolResults") or []:
                if not isinstance(result, dict):
                    continue
                tool_id = result.get("toolCallId") or result.get("toolUseId") or result.get("id") or ""
                tool_name = result.get("toolName") or result.get("name") or ""
                content = result.get("content", result)
                yield from self._emit_tool(tool_id, tool_name, None, content)

    def _handle_message_update(self, obj: dict[str, Any]) -> Generator[dict[str, Any], None, None]:
        ev = obj.get("assistantMessageEvent") or {}
        et = ev.get("type")

        if et == "text_delta":
            delta = ev.get("delta") or ""
            if delta:
                yield from self._emit_text(delta)
            return

        if et == "toolcall_start":
            self._mark_text_sep_needed()
            tool_call = ev.get("toolCall") or {}
            tool_id = tool_call.get("id") or ""
            tool_name = tool_call.get("name") or ""
            if tool_id and tool_id not in self.seen_tool_starts:
                self.seen_tool_starts.add(tool_id)
                yield {"type": "tool_call_start", "toolCallId": tool_id, "toolName": tool_name}
            return

        if et == "toolcall_end":
            tool_call = ev.get("toolCall") or {}
            tool_id = tool_call.get("id") or ""
            tool_name = tool_call.get("name") or ""
            args = tool_call.get("arguments") or tool_call.get("input") or {}
            yield from self._emit_tool(tool_id, tool_name, args, None)

    def _emit_tool(
        self,
        tool_id: str,
        tool_name: str,
        args: Any,
        result: Any,
    ) -> Generator[dict[str, Any], None, None]:
        if not tool_id:
            return

        self._mark_text_sep_needed()

        if tool_id not in self.seen_tool_starts:
            self.seen_tool_starts.add(tool_id)
            yield {"type": "tool_call_start", "toolCallId": tool_id, "toolName": tool_name}

        if args is not None and tool_id not in self.seen_tool_ends:
            args_json = json.dumps(args) if not isinstance(args, str) else args
            yield {"type": "tool_call_args", "toolCallId": tool_id, "toolArgs": args_json}

        if tool_id not in self.seen_tool_ends:
            self.seen_tool_ends.add(tool_id)
            end_args = ""
            if args is not None:
                end_args = json.dumps(args) if not isinstance(args, str) else args
            yield {
                "type": "tool_call_end",
                "toolCallId": tool_id,
                "toolName": tool_name,
                "toolArgs": end_args,
            }

        if result is not None and tool_id not in self.seen_tool_results:
            self.seen_tool_results.add(tool_id)
            content = result if isinstance(result, str) else json.dumps(result)
            yield {"type": "tool_call_result", "toolCallId": tool_id, "content": content}
            yield from _emit_image_events(result)

File: internal/agent/pi_stream.go
// Copyright (c) Mehmet Bektas <mbektasgh@outlook.com>

package agent

import (
	"encoding/json"
	"strings"
)

type piStreamParser struct {
	emittedText     bool
	needsTextSep    bool
	seenToolStarts  map[string]bool
	seenToolEnds    map[string]bool
	seenToolResults map[string]bool
}

func newPiStreamParser() *piStreamParser {
	return &piStreamParser{
		seenToolStarts:  map[string]bool{},
		seenToolEnds:    map[string]bool{},
		seenToolResults: map[string]bool{},
	}
}

func (p *piStreamParser) markTextSepNeeded() {
	if p.emittedText {
		p.needsTextSep = true
	}
}

func (p *piStreamParser) emitText(text string, events chan<- Event) {
	if text == "" {
		return
	}
	if p.emittedText && p.needsTextSep && !strings.HasPrefix(text, "\n") {
		text = "\n\n" + text
	}
	p.needsTextSep = false
	p.emittedText = true
	events <- Event{Type: EventText, Content: text}
}

func (p *piStreamParser) handleLine(line []byte, events chan<- Event) {
	var obj struct {
		Type                 string          `json:"type"`
		ID                   string          `json:"id"`
		ToolCallID           string          `json:"toolCallId"`
		ToolName             string          `json:"toolName"`
		Args                 json.RawMessage `json:"args"`
		Result               json.RawMessage `json:"result"`
		ToolResults          json.RawMessage `json:"toolResults"`
		AssistantMessageEvent json.RawMessage `json:"assistantMessageEvent"`
	}
	if err := json.Unmarshal(line, &obj); err != nil {
		return
	}

	switch obj.Type {
	case "session":
		// session ID is captured via get_state after the turn
	case "message_update":
		p.handleMessageUpdate(obj.AssistantMessageEvent, events)
	case "tool_execution_start":
		p.markTextSepNeeded()
		if obj.ToolCallID != "" && !p.seenToolStarts[obj.ToolCallID] {
			p.seenToolStarts[obj.ToolCallID] = true
			events <- Event{
				Type:       EventToolCallStart,
				ToolCallID: obj.ToolCallID,
				ToolName:   obj.ToolName,
			}
		}
	case "tool_execution_end":
		p.emitTool(obj.ToolCallID, obj.ToolName, obj.Args, obj.Result, events)
	case "turn_end":
		var results []json.RawMessage
		if err := json.Unmarshal(obj.ToolResults, &results); err == nil {
			for _, raw := range results {
				var result struct {
					ToolCallID string          `json:"toolCallId"`
					ToolUseID  string          `json:"toolUseId"`
					ID         string          `json:"id"`
					ToolName   string          `json:"toolName"`
					Name       string          `json:"name"`
					Content    json.RawMessage `json:"content"`
				}
				if json.Unmarshal(raw, &result) != nil {
					continue
				}
				toolID := result.ToolCallID
				if toolID == "" {
					toolID = result.ToolUseID
				}
				if toolID == "" {
					toolID = result.ID
				}
				toolName := result.ToolName
				if toolName == "" {
					toolName = result.Name
				}
				content := result.Content
				if len(content) == 0 {
					content = raw
				}
				p.emitTool(toolID, toolName, nil, content, events)
			}
		}
	}
}

func (p *piStreamParser) handleMessageUpdate(raw json.RawMessage, events chan<- Event) {
	var ev struct {
		Type     string `json:"type"`
		Delta    string `json:"delta"`
		ToolCall struct {
			ID        string          `json:"id"`
			Name      string          `json:"name"`
			Arguments json.RawMessage `json:"arguments"`
			Input     json.RawMessage `json:"input"`
		} `json:"toolCall"`
	}
	if err := json.Unmarshal(raw, &ev); err != nil {
		return
	}

	switch ev.Type {
	case "text_delta":
		if ev.Delta != "" {
			p.emitText(ev.Delta, events)
		}
	case "toolcall_start":
		p.markTextSepNeeded()
		if ev.ToolCall.ID != "" && !p.seenToolStarts[ev.ToolCall.ID] {
			p.seenToolStarts[ev.ToolCall.ID] = true
			events <- Event{
				Type:       EventToolCallStart,
				ToolCallID: ev.ToolCall.ID,
				ToolName:   ev.ToolCall.Name,
			}
		}
	case "toolcall_end":
		args := ev.ToolCall.Arguments
		if len(args) == 0 {
			args = ev.ToolCall.Input
		}
		p.emitTool(ev.ToolCall.ID, ev.ToolCall.Name, args, nil, events)
	}
}

func (p *piStreamParser) emitTool(toolID, toolName string, args, result json.RawMessage, events chan<- Event) {
	if toolID == "" {
		return
	}

	p.markTextSepNeeded()

	if !p.seenToolStarts[toolID] {
		p.seenToolStarts[toolID] = true
		events <- Event{
			Type:       EventToolCallStart,
			ToolCallID: toolID,
			ToolName:   toolName,
		}
	}

	if len(args) > 0 && string(args) != "null" && !p.seenToolEnds[toolID] {
		events <- Event{
			Type:       EventToolCallArgs,
			ToolCallID: toolID,
			ToolArgs:   string(args),
		}
	}

	if !p.seenToolEnds[toolID] {
		p.seenToolEnds[toolID] = true
		endArgs := ""
		if len(args) > 0 && string(args) != "null" {
			endArgs = string(args)
		}
		events <- Event{
			Type:       EventToolCallEnd,
			ToolCallID: toolID,
			ToolName:   toolName,
			ToolArgs:   endArgs,
		}
	}

	if len(result) > 0 && string(result) != "null" && !p.seenToolResults[toolID] {
		p.seenToolResults[toolID] = true
		events <- Event{
			Type:       EventToolCallResult,
			ToolCallID: toolID,
			Content:    string(result),
		}
		emitImageEvents(result, "", events)
	}
}
